src/validators/validation_state.rs

```diff
@@ -53,6 +56,17 @@ impl<'a, 'py> ValidationState<'a, 'py> {
         &self.extra
     }
 
+    pub fn enumerate_last_partial<'i, I>(
+        &self,
+        iter: impl Iterator<Item = I> + 'i,
+    ) -> Box<dyn Iterator<Item = (usize, bool, I)> + 'i> {
+        if self.allow_partial {
+            Box::new(EnumerateLastPartial::new(iter))
+        } else {
+            Box::new(iter.enumerate().map(|(i, x)| (i, false, x)))
+        }
+    }
```

Comment:
Rather than forcing dynamic dispatch here, it might be better to use `.peekable()` at the callsites to be able to check if the current iteration was the last as part of the error pathway.

Reply:
great idea!

Reply:
`peekable` made things worse, but I got rid of the `Box dyn` in 47f1c15

and it helped quite a lot locally - comparing main to 47f1c15:

```
┏━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┓
┃  Group      ┃  Benchmark                          ┃  Before (µs/iter)  ┃  After (µs/iter)  ┃  Change  ┃
┡━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━┩
│  List       │  list_of_ints_core_py               │             29.17  │            30.09  │  +3.17%  │
├─────────────┼─────────────────────────────────────┼────────────────────┼───────────────────┼──────────┤
│  List JSON  │  list_of_ints_core_json             │             42.49  │            43.57  │  +2.55%  │
├─────────────┼─────────────────────────────────────┼────────────────────┼───────────────────┼──────────┤
│  Set        │  set_of_ints_core                   │             52.26  │            52.81  │  +1.06%  │
│             │  set_of_ints_core_duplicates        │             35.59  │            35.51  │  -0.22%  │
│             │  set_of_ints_core_length            │             53.59  │            55.16  │  +2.94%  │
├─────────────┼─────────────────────────────────────┼────────────────────┼───────────────────┼──────────┤
│  Set JSON   │  set_of_ints_core_json              │             56.31  │            58.40  │  +3.70%  │
│             │  set_of_ints_core_json_duplicates   │             44.14  │            45.07  │  +2.12%  │
├─────────────┼─────────────────────────────────────┼────────────────────┼───────────────────┼──────────┤
│  FrozenSet  │  frozenset_of_ints_core             │             16.60  │            17.26  │  +3.99%  │
│             │  frozenset_of_ints_duplicates_core  │             13.78  │            14.78  │  +7.20%  │
├─────────────┼─────────────────────────────────────┼────────────────────┼───────────────────┼──────────┤
│  Dict       │  dict_of_ints_core                  │             85.74  │            86.36  │  +0.73%  │
├─────────────┼─────────────────────────────────────┼────────────────────┼───────────────────┼──────────┤
│  Dict JSON  │  dict_of_ints_core_json             │             127.4  │            126.5  │  -0.72%  │
└─────────────┴─────────────────────────────────────┴────────────────────┴───────────────────┴──────────┘
```